File: KthSmallestElementBST_iterative.py
# TC: O(K) since we return the kth smallest element after traversing the kth element in inorder. 
# SC: O(H) where H is the height of the stack being used.

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        if not root: 
            return
        
        stack = []
        count = 0
        while root or stack: 
            while root:
                stack.append(root)
                root = root.left
            
            curr = stack.pop()
            count += 1
            if count == k: 
                return curr.val
            
            root = curr.right
        
        return -1
        

File: kthSmallestElementBST_Iterative.cs
//https://leetcode.com/submissions/detail/644930946/
// Time: O(n)
//Space: O(n) height of tree.
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    
    public int KthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        
        while(stack.Count > 0 || root !=null ){
            
            while(root != null) {
                stack.Push(root);
                root = root.left;
            }
            
            k--;
            
            root = stack.Pop();
            
            if(k == 0) {
                return root.val;
            }
            
            root = root.right;
        }
         return -1;
    }
    
}